File: 0617-merge-two-binary-trees/0617-merge-two-binary-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        if(!root1 || !root2) return root1 ? root1 : root2;
        function<void(TreeNode*,TreeNode*)> help = [&](TreeNode* root1, TreeNode* root2) -> void {
            root1->val = root1->val + root2->val;
            if(root1->left && root2->left) help(root1->left,root2->left);
            if(root1->right && root2->right) help(root1->right,root2->right);
            if(!root1->left){
                root1->left = root2->left;
                root2->left = nullptr;
            }
            if(!root1->right){
                root1->right = root2->right;
                root2->right = nullptr;
            }
        };
        help(root1,root2);
        return root1;
    }
};

File: 0617-merge-two-binary-trees/0617-merge-two-binary-trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func help(root1 *TreeNode, root2 *TreeNode) {
    root1.Val = root1.Val + root2.Val
    if root1.Left != nil && root2.Left != nil { help(root1.Left,root2.Left) }
    if root1.Right != nil && root2.Right != nil { help(root1.Right,root2.Right) }
    if root1.Left == nil {
        root1.Left = root2.Left
        root2.Left = nil 
    }
    if root1.Right == nil {
        root1.Right = root2.Right
        root2.Right = nil
    }
}

func mergeTrees(root1 *TreeNode, root2 *TreeNode) *TreeNode {
    if root1 == nil || root2 == nil {
        if root1 != nil { return root1 } else { return root2 }
    }
    help(root1,root2)
    return root1
}